src/frontend/js/components/CourseRunEnrollment/index.tsx

```diff
@@ -52,189 +52,155 @@ const messages = defineMessages({
-    },
-    // Determine where to go once initial loading is over based on a list of guards
-    readyInitial: {
-      on: {
-        '': [
-          { target: 'enrolled', cond: 'isEnrolled' },
-          { target: 'closed', cond: 'isCourseRunClosed' },
-          { target: 'idle', cond: 'isLoggedInUser' },
-          // Anonymous is last: only users who are not logged-in but are looking at an
-          // open-for-enrollment course run should end up in this state.
-          { target: 'anonymous' },
-        ],
-      },
-    },
-    anonymous: {},
-    closed: {},
-    enrolled: {},
-    enrollmentLoading: {
-      invoke: {
-        onDone: {
-          target: 'enrolled',
-          actions: assign({ enrollment: (_, event) => event.data }),
-        },
-        onError: { target: 'enrollmentFailed' },
-        src: 'doEnroll',
-      },
-    },
-    enrollmentFailed: {
-      on: {
-        ENROLL: { target: 'enrollmentLoading' },
-      },
-    },
-    idle: {
-      on: {
-        ENROLL: { target: 'enrollmentLoading' },
-      },
-    },
-  },
-});
-
 const headers = {
   'Content-Type': 'application/json',
 };
 
 interface CourseRunEnrollmentProps {
   courseRunId: number;
-  loginUrl: string;
 }
 
+enum Step {
+  ANONYMOUS = 'anonymous',
+  CLOSED = 'closed',
+  LOADING = 'loading',
+  ENROLLED = 'enrolled',
+  ENROLLING = 'enrolling',
+  ENROLLMENT_FAILED = 'enrollmentFailed',
+  FAILED = 'failed',
+  IDLE = 'idle',
+}
+
+enum ActionType {
+  UPDATE_CONTEXT = 'UPDATE_CONTEXT',
+  ENROLL = 'ENROLL',
+  ERROR = 'ERROR',
+}
+interface ReducerState {
+  step: Step;
+  context: {
+    isEnrolled: Maybe<Nullable<Boolean>>;
+    courseRun: Maybe<Nullable<CourseRun>>;
+    currentUser: Maybe<Nullable<User>>;
+  };
+  error?: Error;
+}
+type ReducerAction =
+  | { type: ActionType.UPDATE_CONTEXT; payload: Partial<ReducerState['context']> }
+  | { type: ActionType.ENROLL }
+  | { type: ActionType.ERROR; payload: { error: Error } };
+
+const initialState = {
+  step: Step.LOADING,
+  context: {
+    currentUser: undefined,
+    courseRun: undefined,
+    isEnrolled: undefined,
+  },
+};
+
+const getStepFromContext = (
+  { currentUser, courseRun, isEnrolled }: ReducerState['context'],
+  previousStep: Step,
+) => {
+  switch (true) {
+    case courseRun && courseRun.state.priority > 1:
+      return Step.CLOSED;
+    case previousStep === Step.ENROLLING && !isEnrolled:
+      return Step.ENROLLMENT_FAILED;
+    case currentUser === undefined || courseRun === undefined || isEnrolled === undefined:
+      return Step.LOADING;
+    case currentUser === null:
+      return Step.ANONYMOUS;
+    case !!isEnrolled:
+      return Step.ENROLLED;
+    case !isEnrolled:
+      return Step.IDLE;
+    default:
+      throw new Error('Impossible state');
+  }
+};
+
+const reducer = ({ step, context }: ReducerState, action: ReducerAction): ReducerState => {
+  switch (action.type) {
+    case ActionType.UPDATE_CONTEXT: {
```

Comment:
This is not common to see a `case` with brackets :smiley: can I suggest to remove them ?